### PR TITLE
Avoid ConcurrentHashMap recursive update in ShardingSphereServiceLoader.getServiceInstances

### PR DESCRIPTION
**Pull request by sandynz**
There might be `IllegalStateException: Recursive update` if `ShardingSphereServiceLoader.getServiceInstances` is called recursively. e.g.
```
Caused by: java.lang.IllegalStateException: Recursive update
	at java.base/java.util.concurrent.ConcurrentHashMap.transfer(ConcurrentHashMap.java:2552)
	at java.base/java.util.concurrent.ConcurrentHashMap.addCount(ConcurrentHashMap.java:2354)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1776)
	at org.apache.shardingsphere.infra.util.spi.ShardingSphereServiceLoader.getServiceInstances(ShardingSphereServiceLoader.java:71)
```
After discussion with @TeslaCN , it's better to remove `computeIfAbsent`.

Changes proposed in this pull request:
  - Replace computeIfAbsent in getServiceInstances to avoid possible ConcurrentHashMap recursive update
  - Lazy load service instances to reduce locked time in getServiceInstances

---

Before committing this PR, I'm sure that I have checked the following options:
- [ ] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [ ] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [ ] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

**Comment by sandynz** [comment]
I tested different implementations with JMH
- Replace ConcurrentHashMap to ConcurrentSkipListMap, one line changed (not committed). Performance is bad than before.
- Use ConcurrentHashMap, lazy load (Commit reverted). Performance is a little bad than before.
- Use ConcurrentHashMap, small range locks. Performance is similar as before.

**Comment by sandynz** [comment]
JMH 1.36 test result.

Result:
```
Benchmark                                                    Mode  Cnt    Score    Error   Units
ShardingSphereServiceLoaderJMHTest.loadNewInstanceV1        thrpt    5   23.109 ±  2.709  ops/us
ShardingSphereServiceLoaderJMHTest.loadNewInstanceV2        thrpt    5   24.156 ±  0.694  ops/us
ShardingSphereServiceLoaderJMHTest.loadSingletonInstanceV1  thrpt    5  134.726 ±  4.417  ops/us
ShardingSphereServiceLoaderJMHTest.loadSingletonInstanceV2  thrpt    5  132.867 ± 17.676  ops/us
```

Steps:
1. Copy original code of ShardingSphereServiceLoader.java into ShardingSphereServiceLoader1.java
2. Run ShardingSphereServiceLoaderJMHTest.java to test singleton instances  and new instances loading

Add dependencies in shardingsphere-infra-util module pom.xml:
```
        <dependency>
            <groupId>org.openjdk.jmh</groupId>
            <artifactId>jmh-core</artifactId>
            <version>1.36</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.openjdk.jmh</groupId>
            <artifactId>jmh-generator-annprocess</artifactId>
            <version>1.36</version>
            <scope>test</scope>
        </dependency>
```

Add ShardingSphereServiceLoaderJMHTest.java in `shardingsphere-infra-util` module:
```
package org.apache.shardingsphere.infra.util.spi;

import org.apache.shardingsphere.infra.util.spi.fixture.singleton.SingletonSPIFixture;
import org.apache.shardingsphere.infra.util.spi.type.typed.fixture.TypedSPIFixture;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.RunnerException;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

import java.util.concurrent.TimeUnit;

@BenchmarkMode(Mode.Throughput)
@OutputTimeUnit(TimeUnit.MICROSECONDS)
@State(Scope.Benchmark)
public class ShardingSphereServiceLoaderJMHTest {
    
    public static void main(String[] args) throws RunnerException {
        Options options = new OptionsBuilder().include(ShardingSphereServiceLoaderJMHTest.class.getSimpleName())
                .forks(1).warmupIterations(5).measurementIterations(5)
                .build();
        new Runner(options).run();
    }
    
    @Benchmark
    public void loadSingletonInstanceV1() {
        ShardingSphereServiceLoader1.getServiceInstances(SingletonSPIFixture.class);
    }
    
    @Benchmark
    public void loadSingletonInstanceV2() {
        ShardingSphereServiceLoader.getServiceInstances(SingletonSPIFixture.class);
    }
    
    @Benchmark
    public void loadNewInstanceV1() {
        ShardingSphereServiceLoader1.getServiceInstances(TypedSPIFixture.class);
    }
    
    @Benchmark
    public void loadNewInstanceV2() {
        ShardingSphereServiceLoader.getServiceInstances(TypedSPIFixture.class);
    }
}
```

**Comment by TeslaCN** [comment]
`Recursive update` in ConcurrentHashMap is similar with `java.util.ConcurrentModificationException`. Replacing `compute` methods cannot resolve the root cause.

**Comment by sandynz** [comment]
If users load SPI recursively when customizing SPI, I don't think it's users' fault (root cause).

e.g.
```
public final class OuterSPI implements TypedSPI {

    // TypedSPILoader depends on ShardingSphereServiceLoader
    private final InnerSPI innerSPI = TypedSPILoader.getService(InnerSPI.class, "type");
}
```

Users just use classes of ShardingSphere indirectly, e.g. `ShardingSphereServiceLoader`, but not use `ConcurrentHashMap` directly.
And `ShardingSphereServiceLoader.getServiceInstances` javadoc doesn't declare it could not be invoked recursively.

Maybe we could update javadoc of `ShardingSphereServiceLoader` and also [Dev Manual]( https://shardingsphere.apache.org/document/5.3.1/en/dev-manual/ ).

File: infra/util/src/main/java/org/apache/shardingsphere/infra/util/spi/ShardingSphereServiceLoader.java [hunk 1]
```diff
@@ -70,7 +80,17 @@ private Collection<T> load() {
     @SuppressWarnings("unchecked")
     public static <T> Collection<T> getServiceInstances(final Class<T> serviceInterface) {
         ShardingSphereServiceLoader<?> result = LOADERS.get(serviceInterface);
-        return (Collection<T>) (null != result ? result.getServiceInstances() : LOADERS.computeIfAbsent(serviceInterface, ShardingSphereServiceLoader::new).getServiceInstances());
+        if (null != result) {
+            return (Collection<T>) result.getServiceInstances();
+        }
+        synchronized (LOAD_LOCKS[serviceInterface.hashCode() % LOAD_LOCKS_COUNT]) {
```
Comment (TeslaCN) [inline infra/util/src/main/java/org/apache/shardingsphere/infra/util/spi/ShardingSphereServiceLoader.java:86]: How about `synchronized (serviceInterface)`?

**Comment by TeslaCN** [comment]
> If users load SPI recursively when customizing SPI, I don't think it's users' fault (root cause).
> 
> e.g.
> 
> ```
> public final class OuterSPI implements TypedSPI {
> 
>     // TypedSPILoader depends on ShardingSphereServiceLoader
>     private final InnerSPI innerSPI = TypedSPILoader.getService(InnerSPI.class, "type");
> }
> ```
> 
> Users just use classes of ShardingSphere indirectly, e.g. `ShardingSphereServiceLoader`, but not use `ConcurrentHashMap` directly. And `ShardingSphereServiceLoader.getServiceInstances` javadoc doesn't declare it could not be invoked recursively.
> 
> Maybe we could update javadoc of `ShardingSphereServiceLoader` and also [Dev Manual](https://shardingsphere.apache.org/document/5.3.1/en/dev-manual/).

Got it.

File: infra/util/src/main/java/org/apache/shardingsphere/infra/util/spi/ShardingSphereServiceLoader.java (hunk 1 above)
Comment (sandynz) [reply infra/util/src/main/java/org/apache/shardingsphere/infra/util/spi/ShardingSphereServiceLoader.java:86]: Sorry, I missed this message before.

`serviceInterface` might be synchronized here and in other places, I'm afraid there might be synchronization issue, so I use dedicated lock here.

Comment (TeslaCN) [reply infra/util/src/main/java/org/apache/shardingsphere/infra/util/spi/ShardingSphereServiceLoader.java:86]: Got it.

**Review by TeslaCN — APPROVED** [review]

**Pull request merged**